### PR TITLE
[RPC] decodeblock

### PR DESCRIPTION
**Pull request by instagibbs**
I find this useful for looking at not-PoW-valid blocks rather than using an external parser.

### CODE REVIEW

**Comment by laanwj** [comment]
From a performance perspective using RPC for utility-only functions that could be performed locally, especially something that requires pumping over megabytes of data, is a bad idea.

**Comment by Sjors** [comment]
What about expanding / renaming `getblock` such that it can take either a hash or a full block hex?

**Comment by promag** [comment]
@Sjors please no overloading, especially when he purposes are different.

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -115,16 +115,18 @@ UniValue blockToJSON(const CBlock& block, const CBlockIndex* blockindex, bool tx
 {
     AssertLockHeld(cs_main);
     UniValue result(UniValue::VOBJ);
-    result.pushKV("hash", blockindex->GetBlockHash().GetHex());
+    result.pushKV("hash", block.GetHash().GetHex());
```
Comment (promag) [inline src/rpc/blockchain.cpp]: There is a performance penalty in not using `blockindex->GetBlockHash()`. Suggestion:
```cpp
result.pushKV("hash", blockindex ? blockindex->GetBlockHash().GetHex() : block.GetHash().GetHex());
```

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -769,6 +770,70 @@ static CBlock GetBlockChecked(const CBlockIndex* pblockindex)
...
+        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "Block does not start with a coinbase");
+    }
+
+    int verbosity = 1;
+    if (!request.params[1].isNull()) {
+        verbosity = request.params[1].get_int();
+    }
+
+    return blockToJSON(block, nullptr, verbosity >= 2);
+
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Nit, remove empty line.

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -769,6 +770,70 @@ static CBlock GetBlockChecked(const CBlockIndex* pblockindex)
...
+            "         ,...\n"
+            "  ],\n"
+            "  ,...                     Same output as verbosity = 1.\n"
+            "}\n"
+            "\nExamples:\n"
+            + HelpExampleCli("decodeblock", "\"<block hex>\"")
+            + HelpExampleRpc("decodeblock", "\"<block hex>\"")
+        );
+
+    LOCK(cs_main);
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Could lock after `DecodeHexBlk` below.

**Review by promag — COMMENTED** [review]
Agree with @laanwj above, this is a good candidate for a local tool.

**Comment by laanwj** [comment]
> What about expanding / renaming getblock such that it can take either a hash or a full block hex?

Disagree, if we end up adding this, then adding it as a separate "pure utility" (as in "this function could run without node") RPC call with disparate functionality is best. I think coaxing it into an existing server function unnecessarily complicates the interface.

**Comment by sdaftuar** [comment]
I'm a -0; I agree with the sentiment that this should be done in a standalone tool rather than part of bitcoind.

**Pull request closed without merging**